src/core/types/object/json.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isArray} from '../array';
+
+/**
+ * @fileoverview This module declares JSON types as defined in the
+ * {@link http://json.org/}.
+ */
+
+// NOTE Type are changed to {*} because of
+// https://github.com/google/closure-compiler/issues/1999
+
+/**
+ * JSON scalar. It's either string, number or boolean.
+ * @typedef {string|number|boolean|null}
+ */
+let JSONScalarDef;
+
+/**
+ * JSON object. It's a map with string keys and JSON values.
+ * @typedef {!Object<string, ?*>} (* should be JSONValueDef)
+ */
+let JSONObjectDef;
+
+/**
+ * JSON array. It's an array with JSON values.
+ * @typedef {!Array<?*>} (* should be JSONValueDef)
+ */
+let JSONArrayDef;
+
+/**
+ * JSON value. It's either a scalar, an object or an array.
+ * @typedef {!JSONScalarDef|!JSONObjectDef|!JSONArrayDef}
+ */
+let JSONValueDef;
+
+/**
+ * @typedef {{
+ *   YOU_MUST_USE: string,
+ *   jsonLiteral: function(),
+ *   TO_MAKE_THIS_TYPE: string,
+ * }}
+ */
+let InternalJsonLiteralTypeDef;
+
+/**
+ * Simple wrapper around JSON.parse that casts the return value
+ * to JsonObject.
+ * Create a new wrapper if an array return value is desired.
+ * @param {string} json JSON string to parse
+ * @return {?JsonObject} May be extend to parse arrays.
+ */
+export function parseJson(json) {
+  return /** @type {?JsonObject} */ (JSON.parse(json));
+}
+
+/**
+ * Parses the given `json` string without throwing an exception if not valid.
+ * Returns `undefined` if parsing fails.
+ * Returns the `Object` corresponding to the JSON string when parsing succeeds.
+ * @param {string} json JSON string to parse
+ * @param {function(!Error)=} opt_onFailed Optional function that will be called
+ *     with the error if parsing fails.
+ * @return {?JsonObject} May be extend to parse arrays.
+ */
+export function tryParseJson(json, opt_onFailed) {
+  try {
+    return parseJson(json);
+  } catch (e) {
+    if (opt_onFailed) {
+      opt_onFailed(e);
+    }
```

Comment:
```suggestion
    opt_onFailed?.(e);
```

Reply:
Does not work for `undefined`.

![image](https://user-images.githubusercontent.com/254946/118171694-5aa8f300-b3e0-11eb-9522-25bafe2d7905.png)

Reply:
I had no idea optional chaining worked for function calls. Awesome